### PR TITLE
Locale feature

### PR DESCRIPTION
**Pull request by PauuloG**
Added Locale model and seeded most commonly used locales.

Additions : 
Mutation : `addLocaleToProject`
Queries : `getProjects` `getProject(projectId: ID!)` `getLocales`

Note : we should discuss filtering requested fields to trigger the Objection `eager('relation')` method when loading objects (I think we can find a smart way to query our db here)

### CODE REVIEW

File: src/models/User/createUser/validations/validateEmail/index.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ import validator from 'validator'
 export const INVALID_EMAIL_ERROR =
   'Your e-mail address does not seem to be valid.'
 export const EMAIL_ALREADY_IN_USE_ERROR =
-  'This e-mail address is already in use by another user.'
+  'This e-mail address is already used by another user.'
```
Comment (PauuloG) [inline src/models/User/createUser/validations/validateEmail/index.js]: The grammar warrior rises

File: src/api/projects/queries/getProject/index.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  if (request.user === null) {
+    throw new Error(AUTHENTICATION_ERROR_NO_USER)
+  }
+
+  const project = await Project.query()
+    .eager('locales')
+    .findOne({ id: projectId, ownerId: request.user.id })
+
+  if (!project) {
+    throw new Error('This project was not found')
```
Comment (PauuloG) [inline src/api/projects/queries/getProject/index.js]: Not sure this is the best way to return an error, wdyt ?

File: src/models/Project/addLocaleToProject/index.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+const addLocaleToProject = async ({
+  projectId,
+  localeId,
+  userId
+}: AddLocaleToProjectArgs): Promise<AddProjectToLocaleResponse> => {
+  const errors: Array<ValidationError> = []
+
+  const project = await Project.query()
+    .eager('locales')
```
Comment (PauuloG) [inline src/models/Project/addLocaleToProject/index.js]: This loads the `locales` field while fetching the initial object (otherwise query just returns the object and not its associations).

Maybe we could find a way to filter the eager fields to load only the requested fields (useful if we want to use this in different contexts without pressuring the db)

File: src/models/User/createUser/validations/validateEmail/index.js (hunk 1 above)
Comment (Tom-Bonnike) [reply src/models/User/createUser/validations/validateEmail/index.js]: Is already in use is correct, but if it makes you happy 😄

File: src/api/locales/queries/getLocales/index.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+type Context = {
+  request: express$Request
+}
+
+const resolver = async (_: any, args: Array<any>, { request }: Context) => {
+  if (request.user === null) {
+    throw new Error(AUTHENTICATION_ERROR_NO_USER)
+  }
+
+  const locales = Locale.query()
```
Comment (Tom-Bonnike) [inline src/api/locales/queries/getLocales/index.js]: 🤮

File: src/models/User/createUser/validations/validateEmail/index.js (hunk 1 above)
Comment (PauuloG) [reply src/models/User/createUser/validations/validateEmail/index.js]: Then we should say `This e-mail address is already in use`

Comment (Tom-Bonnike) [reply src/models/User/createUser/validations/validateEmail/index.js]: 🖕

File: src/api/locales/queries/getLocales/index.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// @flow
+import Locale from '@models/Locale'
+import { AUTHENTICATION_ERROR_NO_USER } from '@constants/errors'
+
+type Context = {
+  request: express$Request
+}
+
+const resolver = async (_: any, args: Array<any>, { request }: Context) => {
```
Comment (Tom-Bonnike) [inline src/api/locales/queries/getLocales/index.js]: Args aren’t an array. `args: any` works.
I think if you don’t use them you could also do `_: any` like for the first argument.

File: src/api/locales/queries/getLocales/index.js (hunk 4 above)
Comment (Tom-Bonnike) [reply src/api/locales/queries/getLocales/index.js]: Should we await it?

File: src/api/projects/queries/getProject/index.js (hunk 2 above)
Comment (Tom-Bonnike) [reply src/api/projects/queries/getProject/index.js]: I think it’s fine. Users shouldn’t query projects that don’t exist, so we don’t need to show them a clean error client-side IMO.

That would cause a client-side query to fail, though. But that‘s fine, again it should never happen. :)

File: src/api/projects/queries/getProjects/index.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// @flow
+import Project from '@models/Project'
+import { AUTHENTICATION_ERROR_NO_USER } from '@constants/errors'
+
+type Context = {
+  request: express$Request
+}
+
+const resolver = async (_: any, args: Array<any>, { request }: Context) => {
```
Comment (Tom-Bonnike) [inline src/api/projects/queries/getProjects/index.js]: Same here, `args: any` or `_: any`

File: src/database/migrations/20180517100530_create_locales.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import localesJson from '../seeds/locales.json'
```
Comment (Tom-Bonnike) [inline src/database/migrations/20180517100530_create_locales.js]: I guess we could have a data folder in the migrations and put the locales there? Since it’s not actually a seed, as you pointed out.

File: src/database/migrations/20180517100530_create_locales.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+      .notNull()
+      .defaultTo(knex.fn.now())
+    table.unique(['projectId', 'localeId'])
+  })
+
+  const localesArray = []
+
+  Object.entries(localesJson).map(([key, value]) => {
+    localesArray.push({ code: key, name: value })
+  })
```
Comment (Tom-Bonnike) [inline src/database/migrations/20180517100530_create_locales.js]: You could (should) simply have:

```js
const localesArray = Object.entries(localesJson).map(([key, value]) => ({ code: key, name: value })
```
Otherwise if you’re mutating something else, use `forEach` :)

Comment (Tom-Bonnike) [reply src/database/migrations/20180517100530_create_locales.js]: Also `localesArray`isn’t really the best naming, maybe `localesToSave` or something similar?

File: src/database/migrations/20180517100530_create_locales.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+  await knex('locales').insert(localesArray)
+}
+
+export const down = async knex => {
+  await knex.schema.dropTableIfExists('tokens')
```
Comment (Tom-Bonnike) [inline src/database/migrations/20180517100530_create_locales.js]: We need to fix this down ☝️

File: src/models/Project/addLocaleToProject/index.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    .findOne({ id: projectId, ownerId: userId })
+  const locale = await Locale.query().findById(localeId)
+
+  if (!project || !locale) {
+    const notFound = !project ? 'project' : 'locale'
+    errors.push({
+      field: notFound,
+      message: `This ${notFound} could not be found.`
+    })
```
Comment (Tom-Bonnike) [inline src/models/Project/addLocaleToProject/index.js]: I think we could `throw new Error(…)` here as well? It should never happen. Not sure 😄

File: src/models/Project/addLocaleToProject/index.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    return { errors }
+  }
+
+  project.locales.map(({ id }) => {
```
Comment (Tom-Bonnike) [inline src/models/Project/addLocaleToProject/index.js]: I think we should do a count using the database instead of this map. Also, prefer `forEach` when you’re not using the return value of `.map(…)`.
Or maybe there is a `insertIfNotExists`?

File: src/models/Project/addLocaleToProject/index.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    if (locale.id === id) {
+      errors.push({
+        field: 'locale',
+        message: 'This locale is already activated for this project'
+      })
+
+      return { errors }
+    }
+  })
```
Comment (Tom-Bonnike) [inline src/models/Project/addLocaleToProject/index.js]: Maybe we can drop this whole validation ☝️

File: src/models/Project/addLocaleToProject/index.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  project.locales.map(({ id }) => {
+    if (locale.id === id) {
+      errors.push({
+        field: 'locale',
+        message: 'This locale is already activated for this project'
+      })
+
+      return { errors }
```
Comment (Tom-Bonnike) [inline src/models/Project/addLocaleToProject/index.js]: Same here that’s something that shouldn’t happen client-side so we don’t need an error in the `errors` field? idk if that’s a good way to reason about it 😄

File: src/models/Project/addLocaleToProject/index.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+  try {
+    await Project.query().upsertGraphAndFetch(project, { relate: true })
+
+    return { project, errors }
+  } catch (error) {
+    errors.push({ field: 'generic', message: error.message })
+
+    return { errors }
+  }
```
Comment (Tom-Bonnike) [inline src/models/Project/addLocaleToProject/index.js]: I think we should do this everywhere in another PR.
